src/java/org/apache/cassandra/db/marshal/AbstractCompositeType.java

```diff
@@ -316,6 +326,53 @@ public  <V> void validate(V input, ValueAccessor<V> accessor)
         }
     }
 
+    @Override
+    public void checkConstraints(ByteBuffer input, ColumnConstraints constraints) throws ConstraintViolationException
+    {
+        // no constraints defined for the partition keys
+        if (!constraints.hasRelevantConstraints())
+            return;
+
+        ValueAccessor<ByteBuffer> accessor = ByteBufferAccessor.instance;
+
+        boolean isStatic = readIsStatic(input, accessor);
```

Comment:
The fact that `readIsStatic` also advances / mutates the buffer as a side effect makes me 😢. Not your work on this patch. Half a decade ago no less. :)